mr/do_context.h

```diff
@@ -109,7 +102,20 @@ class RawContext {
     return &ptr->Cast<T>();
   }
 
-  const ShardId& current_shard() const { return current_shard_;}
+  // Sometimes we run 2 shards per thread, in which case it is important to have these per-fiber
+  struct PerFiber {
+    std::string file_name;
+    ShardId current_shard;
+    InputMetaData metadata;
+    size_t input_pos = 0;
+    bool is_binary = false;
+  };
+
+  const PerFiber *per_fiber() {
```

Comment:
Looks good except that i prefer having just a pure accessor here. I think it's better to avoid lazy initialization if possible. you have a convenient init point in `MapperExecutor::MapFiber` where you can initialize per fiber structure.